braindecode/models/eegconformer.py

```diff
@@ -0,0 +1,350 @@
+                          "more channels.", UserWarning)
+
+        self.patch_embedding = _PatchEmbedding(
+                n_filters_time=n_filters_time,
+                filter_time_length=filter_time_length,
+                n_channels=n_channels,
+                pool_time_length=pool_time_length,
+                stride_avg_pool=pool_time_stride,
+                drop_prob=drop_prob)
+
+        self.transformer = _TransformerEncoder(
+            att_depth=att_depth,
+            emb_size=n_filters_time,
+            att_heads=att_heads,
+            att_drop=att_drop_prob)
+
+        self.classification_head = _ClassificationHead(
+            emb_size=n_filters_time, final_fc_length=final_fc_length,
+            n_classes=n_classes, return_features=return_features)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = torch.unsqueeze(x, dim=1)  # add one extra dimension
+        x = self.patch_embedding(x)
+        x = self.transformer(x)
+        x = self.classification_head(x)
+        return x
+
+
+class _PatchEmbedding(nn.Module):
+    """Patch Embedding.
+
+    The authors used a convolution module to capture local features,
+    instead of position embedding.
+
+    Parameters
+    ----------
+    n_filters_time: int
+        Number of temporal filters, defines also embedding size.
+    filter_time_length: int
+        Length of the temporal filter.
+    n_channels: int
+        Number of channels to be used as number of spatial filters.
+    pool_time_length: int
+        Length of temporal poling filter.
+    stride_avg_pool: int
+        Length of stride between temporal pooling filters.
+    drop_prob: float
+        Dropout rate of the convolutional layer.
+
+    Returns
+    -------
+    x: torch.Tensor
+        The output tensor of the patch embedding layer.
+    """
+
+    def __init__(
+            self,
+            n_filters_time,
+            filter_time_length,
+            n_channels,
+            pool_time_length,
+            stride_avg_pool,
+            drop_prob,
+    ):
+        super().__init__()
+
+        self.shallownet = nn.Sequential(
+            nn.Conv2d(1, n_filters_time,
+                      (1, filter_time_length), (1, 1)),
+            nn.Conv2d(n_filters_time, n_filters_time,
+                      (n_channels, 1), (1, 1)),
+
+            nn.BatchNorm2d(num_features=n_filters_time),
+            nn.ELU(),
+            nn.AvgPool2d(
+                kernel_size=(1, pool_time_length),
+                stride=(1, stride_avg_pool)
+            ),
+            # pooling acts as slicing to obtain 'patch' along the
+            # time dimension as in ViT
+            nn.Dropout(p=drop_prob),
+        )
+
+        self.projection = nn.Sequential(
+            nn.Conv2d(
+                n_filters_time, n_filters_time, (1, 1), stride=(1, 1)
+            ),  # transpose, conv could enhance fiting ability slightly
+            Rearrange("b d_model 1 seq -> b seq d_model"),
+        )
```

Comment:
Possible extension: additional parameter `embedding_size`. Otherwise the projection layer is not necessary.